server/src/main/java/org/elasticsearch/search/aggregations/TopBucketBuilder.java

```diff
@@ -183,6 +187,8 @@ public void add(DelayedBucket<B> bucket) {
                 next.add(bucket);
                 return;
             }
+            // add one bucket to the final result
+            reduceContext.consumeBucketsAndMaybeBreak(1);
```

Comment:
This should avoid the buffer to grow to ridiculous sizes if the required size provided by the user is ridiculous big.